### PR TITLE
Allow the Tick operation to be either permissioned or permissionless

### LINKED ISSUE #217: Allow IBC Forwarder to be ticked directly
It should be possible to tick the IBC Forwarder directly, not just from the Clock contract.

### PR DESCRIPTION
**Pull request by stiiifff**
Allow the Tick operation to be either permissioned or permissionless based on a set of privileged addresses.
Removed the dependency to the Valence clock .. still something to be addressed (see below in comments).
Cleaned up IBC Forwarder tests related to testing the fallback address (to use a test address instead of reusing clock/next contract).

closes #217

### CODE REVIEW

**Comment by stiiifff** [comment]
@bekauz Open a draft PR so you can already take a look. /cc @uditvira 

Some points left to be addressed:
- I removed the dependency to the Valence clock. One problem thus is that the IBC Forwarder does not enqueue itself into the Clock anymore. Several options:
  - **Leave it as is**: the forwarder can be configured with the privileged_addresses param to allow ticking by the Covenant admin, or to be permissionless (anyone).
  - **Bring back the dependency**: just so that the contract can enqueue itself itself to the Clock (problem we don't have a concept of clock anymore in the forwarder .. which I think is best).
  - **Modify the Clock contract**: allow to pass a list of contract addresses in the Clock instantiate message to pre-initialize the clock Queue with one or more contract(s) (actually we already pass the list of contracts for the whitelist .. so it wouldn't be a big change).

Remaining to do:
- Add tests to verify permissionless ticking, and permissioned ticking with more than one privileged address (e.g. Clock & Covenant admin).

**Comment by stiiifff** [comment]
Commit 6f8c91c962c35651139eb9d8eeb37a7805d8eea1 adds an `initial_queue` instantiation param to the Clock which should forego the need of contracts to enqueue themselves into the Clock. (option 3 above)

**Comment by stiiifff** [comment]
Remaining to do:
- Add tests to verify permissionless ticking, and permissioned ticking with more than one privileged address (e.g. Clock & Covenant admin).

**Review by bekauz — COMMENTED** [review]
looks great! just leaving a couple of (somewhat hypothetical) thoughts here, along with a few code comments.

given that in top level covenants we no longer add the ibc-forwarders to the whitelist, we could run into a situation where a forwarder is instantiated as part of a covenant, does something, and gets dequeued. now, because ibc forwarder was enqueued while bypassing the whitelist, we would not be able to re-queue it easily (e.g. firing `enqueue_clock` message) without update-migrating the clock's whitelist/initial queue state variables.

some ideas that come to mind:
- continue adding the forwarders to the whitelist while also doing the `initial_queue`
- automatically whitelist the initial queue contracts on clock instantiation
- skip the `initial_queue` altogether. "fire and forget"  `enqueue_clock` messages to every privileged address on ibc forwarder instantiation

what do you think?

File: contracts/ibc-forwarder/src/helpers.rs [hunk 1]
```diff
@@ -81,3 +83,31 @@ pub struct MsgTransfer {
     #[prost(string, tag = "8")]
     pub memo: String,
 }
+
+pub fn validate_privileged_addresses(
+    api: &dyn Api,
+    privileged_addresses: Option<Vec<String>>,
+) -> Result<Option<Vec<Addr>>, StdError> {
```
[the remark this answers is not in the document]
Comment (stiiifff) [reply contracts/ibc-forwarder/src/helpers.rs:90]: Fair point. Though I assume that _privileged_adresses_ will typically contain 1 or 2 address(es) so not much of a big deal.

File: unit-tests/src/test_ibc_forwarder/suite.rs [hunk 2]
```diff
@@ -78,6 +79,12 @@ impl IbcForwarderBuilder {
         self
     }
 
+    pub fn with_test_fallback_address(mut self) -> Self {
```
[the remark this answers is not in the document]
Comment (stiiifff) [reply unit-tests/src/test_ibc_forwarder/suite.rs:82]: Yeah good point, I'll make the change.

**Comment by stiiifff** [comment]
I guess that would work. But why should contracts enqueue themselves to another contract .. while the top-level covenant has all the info to act as orchestrator, and setup everything correctly (cfr. whitelist & initial_queue).

**Comment by uditvira** [comment]
Some thoughts:
- I think each of the contracts should have no notion of the clock. Enqueue logic if any should be handled only by the orchestrator (i.e., top level covenant or a human)
- A nice property here is that this will allow the clock to be instantiated by the top level covenant however we can still set the privileged addresses to be empty. This gives us the benefit of ticking any of the contracts directly or ticking the clock, both should work
- In discussion with @stiiifff, we talked about how the top level contract could expose a configuration option enabling/disabling permissionless ticking (i.e., whether the clock is in the privileged addresses or not)

**Review by uditvira — DISMISSED** [review]

**Comment by stiiifff** [comment]
Will add a few more tests before merging, and create separate issues to migrate other contracts to use the same approach (privileged addresses), and to add an option to the top-level covenant to configure permissioned / permissionless ticking.

**Review by uditvira — DISMISSED** [review]
Thanks @stiiifff for adding the additional tests

File: contracts/ibc-forwarder/src/helpers.rs [hunk 3]
```diff
@@ -81,3 +83,29 @@ pub struct MsgTransfer {
...
```
Comment (bekauz) [inline contracts/ibc-forwarder/src/helpers.rs:90]: why vector if we only do `.contains()`?

**Review by bekauz — COMMENTED** [review]
left a few more comments, mostly naming.

personally I try to avoid using types like `Option<Option<Vec>>`, but I can see why you used it that way in the migrate msg. I think whether it can be boiled down to just `Option<Vec>` will depend on what is the expected behaviour when contract is initialized with no permitted caller addresses?

File: contracts/ibc-forwarder/src/helpers.rs (hunk 3 above)
Comment (stiiifff) [reply contracts/ibc-forwarder/src/helpers.rs:90]: done

File: contracts/ibc-forwarder/src/contract.rs [hunk 4]
```diff
@@ -73,7 +79,6 @@ pub fn instantiate(
     }
 
     Ok(Response::default()
```
[the remark this answers is not in the document]
Comment (stiiifff) [reply contracts/ibc-forwarder/src/contract.rs:81]: done

File: unit-tests/src/test_ibc_forwarder/tests.rs [hunk 5]
```diff
@@ -17,17 +17,29 @@ fn test_instantiate_validates_next_contract_addr() {
         .build();
 }
```
[the remark this answers is not in the document]
Comment (stiiifff) [reply unit-tests/src/test_ibc_forwarder/tests.rs:19]: done

File: contracts/ibc-forwarder/src/state.rs [hunk 6]
```diff
@@ -10,8 +10,7 @@ use crate::msg::ContractState;
 /// tracks the current state of state machine
 pub const CONTRACT_STATE: Item<ContractState> = Item::new("contract_state");
 
-/// clock module address to verify the sender of incoming ticks
-pub const CLOCK_ADDRESS: Item<Addr> = Item::new("clock_address");
+pub const PRIVILEGED_ADDRESSES: Item<Option<Vec<Addr>>> = Item::new("privileged_addresses");
```
[the remark this answers is not in the document]
Comment (stiiifff) [reply contracts/ibc-forwarder/src/state.rs]: Authorized usually comes with a definition of roles and/or permissions in a role-based access control system. I think privileged is more generic, and widely used. That being said, I thought _privileged_addresses_ was a bit of a mouthful (plus we don't really care about the addresses themselves), so I've renamed it to _privileged_accounts_ which I think is better, wdyt ?

**Review by bekauz — APPROVED** [review]

**Comment by stiiifff** [comment]
LGTM !

**Pull request merged**